Angular v5 ngc compiler: Error encountered in metadata generated for exported symbol 'Subscription'

<!--
PLEASE HELP US PROCESS GITHUB ISSUES FASTER BY PROVIDING THE FOLLOWING INFORMATION.

ISSUES MISSING IMPORTANT INFORMATION MAY BE CLOSED WITHOUT INVESTIGATION.
-->

## I'm submitting a...
<!-- Check one of the following options with "x" -->
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[X ] Bug report  <!-- Please search GitHub for a similar issue or PR before submitting -->
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
Building an Angulary library using _5.0.0-beta.4 ngc compiler_, I'm getting the following error:
```
Error: Error: C:/Users/Roberto/Documents/GitHub/angular-library-starter/node_modules/rxjs/Subscription.d.ts:21:1: Error encountered in met
adata generated for exported symbol 'Subscription':
 C:/Users/Roberto/Documents/GitHub/angular-library-starter/node_modules/rxjs/Subscription.d.ts:22:12: Metadata collected contains an error
 that will be reported at runtime: Only initialized variables and constants can be referenced because the value of this variable is needed
 by the template compiler.
  {"__symbolic":"error","message":"Variable not initialized","line":21,"character":11}
    at C:\Users\Roberto\Documents\GitHub\angular-library-starter\node_modules\@angular\tsc-wrapped\src\collector.js:658:27
    at Array.forEach (native)
    at validateMetadata (C:\Users\Roberto\Documents\GitHub\angular-library-starter\node_modules\@angular\tsc-wrapped\src\collector.js:646:
42)
    at MetadataCollector.getMetadata (C:\Users\Roberto\Documents\GitHub\angular-library-starter\node_modules\@angular\tsc-wrapped\src\coll
ector.js:503:17)
    at LowerMetadataCache.getMetadataAndRequests (C:\Users\Roberto\Documents\GitHub\angular-library-starter\node_modules\@angular\compiler
-cli\src\transformers\lower_expressions.js:169:39)
    at LowerMetadataCache.ensureMetadataAndRequests (C:\Users\Roberto\Documents\GitHub\angular-library-starter\node_modules\@angular\compi
ler-cli\src\transformers\lower_expressions.js:147:27)
    at LowerMetadataCache.getMetadata (C:\Users\Roberto\Documents\GitHub\angular-library-starter\node_modules\@angular\compiler-cli\src\tr
ansformers\lower_expressions.js:139:21)
    at CompilerHost.upgradeVersion1Metadata (C:\Users\Roberto\Documents\GitHub\angular-library-starter\node_modules\@angular\compiler-cli\
src\compiler_host.js:236:45)
    at CompilerHost.getMetadataFor (C:\Users\Roberto\Documents\GitHub\angular-library-starter\node_modules\@angular\compiler-cli\src\compi
ler_host.js:195:30)
    at StaticSymbolResolver.getModuleMetadata (C:\Users\Roberto\Documents\GitHub\angular-library-starter\node_modules\@angular\compiler\bu
ndles\compiler.umd.js:25829:62)
```

## Minimal reproduction of the problem with instructions
Download the following repo: https://github.com/robisim74/angular-library-starter

Upgrade to Angular 5.0.0-beta.4 and run `npm run build`.

## What is the motivation / use case for changing the behavior?
_Up to Angular 5.0.0-beta.3 it worked without problems. The library does not use itself `rxjs/Subscription`._

The only way to avoid the error seems to be to set `strictMetadataEmit` to _false_, but it is not recommended.


## Environment

<pre><code>
Angular version: 5.0.0-beta.4
<!-- Check whether this is still an issue in the most recent Angular version -->
</code></pre>

Thanks.

--- a/packages/compiler-cli/src/transformers/lower_expressions.ts
+++ b/packages/compiler-cli/src/transformers/lower_expressions.ts
@@ -322,8 +322,14 @@ export class LowerMetadataCache implements RequestsMap {
       return value;
     };
 
+    // Do not validate or lower metadata in a declaration file. Declaration files are requested
+    // when we need to update the version of the metadata to add informatoin that might be missing
+    // in the out-of-date version that can be recovered from the .d.ts file.
+    const declarationFile = sourceFile.isDeclarationFile;
+
     const metadata = this.collector.getMetadata(
-        sourceFile, this.strict, sourceFile.isDeclarationFile ? undefined : substituteExpression);
+        sourceFile, this.strict && !declarationFile,
+        declarationFile ? undefined : substituteExpression);
 
     return {metadata, requests};
   }

--- a/packages/compiler-cli/test/transformers/lower_expressions_spec.ts
+++ b/packages/compiler-cli/test/transformers/lower_expressions_spec.ts
@@ -98,6 +98,38 @@ describe('Expression lowering', () => {
       expect(collected.requests.has(collected.annotations[0].start))
           .toBeTruthy('did not find the data field');
     });
+
+    it('should throw a validation execption for invalid files', () => {
+      const cache = new LowerMetadataCache({}, /* strict */ true);
+      const sourceFile = ts.createSourceFile(
+          'foo.ts', `
+        import {Injectable} from '@angular/core';
+
+        class SomeLocalClass {}
+        @Injectable()
+        export class SomeClass {
+          constructor(a: SomeLocalClass) {}
+        }
+      `,
+          ts.ScriptTarget.Latest, true);
+      expect(() => cache.getMetadata(sourceFile)).toThrow();
+    });
+
+    it('should not report validation errors on a .d.ts file', () => {
+      const cache = new LowerMetadataCache({}, /* strict */ true);
+      const dtsFile = ts.createSourceFile(
+          'foo.d.ts', `
+        import {Injectable} from '@angular/core';
+
+        class SomeLocalClass {}
+        @Injectable()
+        export class SomeClass {
+          constructor(a: SomeLocalClass) {}
+        }
+      `,
+          ts.ScriptTarget.Latest, true);
+      expect(() => cache.getMetadata(dtsFile)).not.toThrow();
+    });
   });
 });
 

